src/ray/object_manager/object_manager.cc

```diff
@@ -196,19 +207,48 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
   return ray::Status::OK();
 }
 
-ray::Status ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id,
-                                int retry) {
+void ObjectManager::HandlePushTaskTimeout(const ObjectID &object_id,
+                                          const ClientID &client_id) {
+  RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
+                 << " after waiting for " << config_.push_timeout_ms << " ms.";
+  auto iter = unfulfilled_push_tasks_.find(object_id);
+  if (iter != unfulfilled_push_tasks_.end()) {
+    iter->second.erase(client_id);
+    if (iter->second.size() == 0) {
+      unfulfilled_push_tasks_.erase(iter);
+    }
+  }
+}
+
+ray::Status ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id) {
   if (local_objects_.count(object_id) == 0) {
-    if (retry < 0) {
-      retry = config_.max_push_retries;
-    } else if (retry == 0) {
-      RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
-                     << " after retrying " << config_.max_push_retries << " times.";
-      return ray::Status::OK();
+    // Avoid setting duplicated timer for the same object and client pair.
+    auto &clients = unfulfilled_push_tasks_[object_id];
+    if (clients.count(client_id) == 0) {
+      // If config_.push_timeout_ms < 0, we give an empty timer
+      // and the task will be kept infinitely.
```

Comment:
It might make sense to add tests for these special configuration settings. Since the current tests do not support changing configuration, this is optional:
1. `config_.push_timeout_ms < 0`; A regular `Push` with this setting is sufficient.
2. `config_.push_timeout_ms == 0`; Do a regular `Push` and a `Push` with an object that doesn't exist.

If you don't have time to do this, I will eventually update the tests to support this and add this test at a later time.

Reply:
I have finished adding tests to cover the code paths.